### PR TITLE
[SPARK-46862][SQL] Disable CSV column pruning in the multi-line mode

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to disable the column pruning feature in the CSV datasource for the `multiLine` mode.

### Why are the changes needed?
To workaround the issue in the `uniVocity` parser used by the CSV datasource: https://github.com/uniVocity/univocity-parsers/issues/529

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running the affected test suites:
```
$ build/sbt "test:testOnly *CSVv1Suite"
$ build/sbt "test:testOnly *CSVv2Suite"
$ build/sbt "test:testOnly *CSVLegacyTimeParserSuite"
$ build/sbt "testOnly *.CsvFunctionsSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
do we have a test to show the data correctness issue?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 1]
```diff
@@ -3212,6 +3215,26 @@ abstract class CSVSuite
...
+        |"8","SE","","No"""".stripMargin
+
+    withTempPath { path =>
+      Files.write(path.toPath, data.getBytes(StandardCharsets.UTF_8))
+      val df = spark.read
+        .option("multiline", "true")
+        .option("header", "true")
+        .option("escape", "\"")
+        .csv(path.getCanonicalPath)
+      assert(df.count() === 5)
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3235]: `count()` returns 4 without the changes.

**Comment by MaxGekk** [comment]
> do we have a test to show the data correctness issue?

@cloud-fan I added a test for the issue.

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon FYI, since 3.5.x and 3.4.x suffer from the same issue, I am going to backport this the branches. Thanks for review.

**Comment by MaxGekk** [comment]
Merging to master/3.5/3.4. Thank you, @HyukjinKwon @cloud-fan for review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala [hunk 2]
```diff
@@ -58,7 +58,7 @@ case class CSVPartitionReaderFactory(
       actualReadDataSchema,
       options,
       filters)
-    val schema = if (options.columnPruning) actualReadDataSchema else actualDataSchema
+    val schema = if (options.isColumnPruningEnabled) actualReadDataSchema else actualDataSchema
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala:61]: https://github.com/apache/spark/blob/829e742df8251c6f5e965cb08ad454ac3ee1a389/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala#L103

https://github.com/apache/spark/blob/829e742df8251c6f5e965cb08ad454ac3ee1a389/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala#L128

@MaxGekk Should the check in `CSVFileFormat` be changed to `parsedOptions.isColumnPruningEnabled` too?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 3]
```diff
@@ -3212,6 +3215,26 @@ abstract class CSVSuite
...
+        |"1","DE","","Yes"
+        |"5",",","",","
+        |"3","SA","","No"
+        |"10","abcd""efgh"" \ndef","",""
+        |"8","SE","","No"""".stripMargin
+
+    withTempPath { path =>
+      Files.write(path.toPath, data.getBytes(StandardCharsets.UTF_8))
+      val df = spark.read
+        .option("multiline", "true")
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3231]: `.option("multiLine", "true")`?

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3231]: the options are case-insensitive

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala:61]: The `schema` is used only `CSVHeaderChecker` which is supposed to check column names in CSV and provided schema fields. It shouldn't depend on the column pruning feature at all, from my point of view.

```scala
  private def checkHeaderColumnNames(columnNames: Array[String]): Unit = {
...
      if (headerLen == schemaSize) {
...
      } else {
        errorMessage = Some(
          s"""|Number of column in CSV header is not equal to number of fields in the schema:
              | Header length: $headerLen, schema size: $schemaSize
              |$source""".stripMargin)
      }
```

`schemaSize` must be f**ull data schema** of CSV filed, but not the required schema.

Let me re-think it, and avoid the dependency from the column pruning in `CSVHeaderChecker`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 3 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3231]: Thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala:61]: @LuciferYang Actually, you are right. Please, review this follow up PR: https://github.com/apache/spark/pull/44910

**Comment by nchammas** [comment]
> To workaround the issue in the `uniVocity` parser used by the CSV datasource: [uniVocity/univocity-parsers#529](https://github.com/uniVocity/univocity-parsers/issues/529)

A bit off-topic for this PR, but is uniVocity even maintained anymore?

- The last release was [more than 3 years ago][1].
- The last commit to `master` was [almost 3 years ago][2].
- The website is [down][3].
- There are [multiple][4] [open][5] [bugs][6] on the tracker with no indication that anyone cares.

[1]: https://github.com/uniVocity/univocity-parsers/releases
[2]: https://github.com/uniVocity/univocity-parsers/commits/master/
[3]: https://github.com/uniVocity/univocity-parsers/issues/506
[4]: https://github.com/uniVocity/univocity-parsers/issues/494
[5]: https://github.com/uniVocity/univocity-parsers/issues/495
[6]: https://github.com/uniVocity/univocity-parsers/issues/499

**Comment by cloud-fan** [comment]
is there any other popular Java libraries for parsing CSV?

**Comment by nchammas** [comment]
I've filed SPARK-47180 to track potentially migrating off of Univocity to something else.

**Pull request closed without merging**